src/main/java/com/exactpro/th2/healer/grpc/HealerServiceImpl.java

```diff
@@ -0,0 +1,185 @@
+
+    @Override
+    public void sendEvent(EventDataRequest request, StreamObserver<EventResponse> responseObserver) {
+        try {
+            LOGGER.info("sendEvent request: {}", request);
+
+            if (!knownCrawlers.contains(request.getId())) {
+                LOGGER.warn("Received request from unknown crawler with id {}. Sending response with HandshakeRequired = true", toJson(request.getId()));
+                responseObserver.onNext(EventResponse.newBuilder()
+                        .setStatus(Status.newBuilder().setHandshakeRequired(true))
+                        .build());
+                responseObserver.onCompleted();
+                return;
+            }
+
+            int eventsCount = request.getEventDataCount();
+
+            heal(request.getEventDataList());
+
+            EventID lastEventId = null;
+
+            if (eventsCount > 0) {
+                lastEventId = request.getEventDataList().get(eventsCount - 1).getEventId();
+            }
+
+            EventResponse.Builder builder = EventResponse.newBuilder();
+
+            if (lastEventId != null) {
+                builder.setId(lastEventId);
+            }
+
+            EventResponse response = builder.build();
+
+            LOGGER.info("sendEvent response: {}", response);
+            responseObserver.onNext(response);
+            responseObserver.onCompleted();
+        } catch (Exception e) {
+            responseObserver.onError(e);
+            LOGGER.error("sendEvent error: " + e, e);
+        }
+    }
+
+    private void heal(Collection<EventData> events) throws IOException {
+        List<InnerEvent> eventAncestors;
+
+        for (EventData event: events) {
+            if (event.getSuccessful() == EventStatus.FAILED && event.hasParentEventId()) {
+
+                eventAncestors = getAncestors(event);
+
+                for (InnerEvent ancestor : eventAncestors) {
+                    StoredTestEventWrapper ancestorEvent = ancestor.event;
+
+                    if (ancestor.success) {
+                        storage.updateEventStatus(ancestorEvent, false);
+                        ancestor.markFailed();
+                        LOGGER.info("Event {} healed", ancestorEvent.getId());
+                    }
+                }
+
+            }
+        }
+    }
+
+    private List<InnerEvent> getAncestors(EventData event) throws IOException {
+        List<InnerEvent> eventAncestors = new ArrayList<>();
+        String parentId = event.getParentEventId().getId();
+
+        while (parentId != null) {
+            InnerEvent innerEvent;
+
+            if (cache.containsKey(parentId)) {
+                innerEvent = cache.get(parentId);
+            } else {
+                StoredTestEventWrapper parent = storage.getTestEvent(new StoredTestEventId(parentId));
+
+                innerEvent = new InnerEvent(parent, parent.isSuccess());
+                cache.put(parentId, innerEvent);
+            }
+
+            eventAncestors.add(innerEvent);
```

Comment:
I am not sure but I think we can stop collecting parents if one of the parents is already failed. That means all its parents should be healed already